src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -0,0 +1,1503 @@
+                        }
+
+                        WriteCompressedUInt((uint)value.Length);
+                        _writer.Write(bytes);
+                    }
+                }
+            }
+        }
+
+        private void WriteBoxedEnum(object value, Type enumType)
+        {
+            _writer.Write((byte)EncodingKind.Enum);
+            this.WriteType(enumType);
+
+            var type = Enum.GetUnderlyingType(enumType);
+
+            if (type == typeof(int))
+            {
+                _writer.Write((int)value);
+            }
+            else if (type == typeof(short))
+            {
+                _writer.Write((short)value);
+            }
+            else if (type == typeof(byte))
+            {
+                _writer.Write((byte)value);
+            }
+            else if (type == typeof(long))
+            {
+                _writer.Write((long)value);
+            }
+            else if (type == typeof(sbyte))
+            {
+                _writer.Write((sbyte)value);
+            }
+            else if (type == typeof(ushort))
+            {
+                _writer.Write((ushort)value);
+            }
+            else if (type == typeof(uint))
+            {
+                _writer.Write((uint)value);
+            }
+            else if (type == typeof(ulong))
+            {
+                _writer.Write((ulong)value);
+            }
+            else
+            {
+                throw ExceptionUtilities.UnexpectedValue(type);
+            }
+        }
+
+        private void WriteArray(Array array)
+        {
+            var elementType = array.GetType().GetElementType();
+
+            EncodingKind elementKind;
+            if (s_typeMap.TryGetValue(elementType, out elementKind))
+            {
+                int length = array.GetLength(0);
+
+                switch (length)
+                {
+                    case 0:
+                        _writer.Write((byte)EncodingKind.Array_0);
+                        break;
+                    case 1:
+                        _writer.Write((byte)EncodingKind.Array_1);
+                        break;
+                    case 2:
+                        _writer.Write((byte)EncodingKind.Array_2);
+                        break;
+                    case 3:
+                        _writer.Write((byte)EncodingKind.Array_3);
+                        break;
+                    default:
+                        _writer.Write((byte)EncodingKind.Array);
+                        this.WriteCompressedUInt((uint)length);
+                        break;
+                }
+
+                this.WritePrimitiveType(elementType, elementKind);
+                this.WritePrimitiveTypeArrayElements(elementType, elementKind, array);
+            }
+            else
+            {
+                // gather list elements
+                _memberWriter.List.Clear();
+
+                foreach (var element in array)
```

Comment:
so this actually copy same data twice? from original array to variant list array and then from variant array to stack?